cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,406 @@
+   * @param comm initialized nccl comm
+   * @param ucp_worker initialized ucp_worker instance
+   * @param eps shared pointer to array of ucp endpoints
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, ucp_worker_h ucp_worker,
+            std::shared_ptr<ucp_ep_h *> eps, int size, int rank)
+    : _nccl_comm(comm),
+      _ucp_worker(ucp_worker),
+      _ucp_eps(eps),
+      _size(size),
+      _rank(rank),
+      _next_request_id(0) {
+    initialize();
+    p2p_enabled = true;
+  };
+
+  /**
+   * @brief constructor for collective-only operation
+   * @param comm initilized nccl communicator
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, int size, int rank)
+    : _nccl_comm(comm), _size(size), _rank(rank) {
+    initialize();
+  };
+
+  virtual ~std_comms() {
+    CUDA_CHECK_NO_THROW(cudaStreamDestroy(_stream));
+
+    CUDA_CHECK_NO_THROW(cudaFree(_sendbuff));
+    CUDA_CHECK_NO_THROW(cudaFree(_recvbuff));
+  }
+
+  void initialize() {
+    CUDA_CHECK(cudaStreamCreate(&_stream));
+
+    CUDA_CHECK(cudaMalloc(&_sendbuff, sizeof(int)));
+    CUDA_CHECK(cudaMalloc(&_recvbuff, sizeof(int)));
+  }
+
+  int getSize() const { return _size; }
+
+  int getRank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> commSplit(int color, int key) const {
+    // Not supported by NCCL
+    ASSERT(false,
+           "ERROR: commSplit called but not yet supported in this comms "
+           "implementation.");
+  }
+
+  void barrier() const {
+    CUDA_CHECK(cudaMemsetAsync(_sendbuff, 1, sizeof(int), _stream));
+    CUDA_CHECK(cudaMemsetAsync(_recvbuff, 1, sizeof(int), _stream));
+
+    allreduce(_sendbuff, _recvbuff, 1, INT, SUM, _stream);
+
+    ASSERT(syncStream(_stream) == status_t::commStatusSuccess,
+           "ERROR: syncStream failed. This can be caused by a failed rank.");
+  }
+
+  void get_request_id(request_t *req) const {
+    request_t req_id;
+
+    if (this->_free_requests.empty())
+      req_id = this->_next_request_id++;
+    else {
+      auto it = this->_free_requests.begin();
+      req_id = *it;
+      this->_free_requests.erase(it);
+    }
+    *req = req_id;
+  }
+
+  void isend(const void *buf, int size, int dest, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
+           "ERROR: UCX comms not initialized on communicator.");
+
+    get_request_id(request);
+    ucp_ep_h ep_ptr = (*_ucp_eps)[dest];
+
+    ucp_request *ucp_req = (ucp_request *)malloc(sizeof(ucp_request));
+
+    this->_ucp_handler.ucp_isend(ucp_req, ep_ptr, buf, size, tag,
+                                 default_tag_mask, getRank());
+
+    _requests_in_flight.insert(std::make_pair(*request, ucp_req));
+  }
+
+  void irecv(void *buf, int size, int source, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
```

Comment:
I like this. Where do you think this should live?

Reply:
These are copy-and-paste-and-modify of https://github.com/rapidsai/cudf/blob/branch-0.14/cpp/include/cudf/utilities/error.hpp, and I think eventually, all these should go to rapids-core.

At this moment, just follow cudf, and place under include/raft/error.hpp or include/raft/utilities/error.hpp?

Reply:
And another question is whether we should place CUML_EXPECTS, CUML_FAIL, CUGRAPH_EXPECTS, CUGRAPH_FAIL in raft or cuml & cudf. If we place those here, it will be helpful in enforcing consistent error handling mechanism and minimizing code duplicates.

Reply:
Yes, I was suggesting having a `RAFT_EXPECT` and other error macros here (like `NCCL_TRY`) so that we can use this consistently across RAFT, cuML and cuGraph repos. 
I don't think rapids-core is happening anytime soon so better move forward with this now through RAFT.

Reply:
I have a concern with this. The current *_EXPECT does not seem to allow formatting of string to provide more meaningful error messages. For eg: We should allow devs to do something like this:
` RAFT_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch rhs=%s lhs=%s", to_string(lhs->dtype), to_string(rhs->dtype));`

I know that core-C++ folks might cringe at C-style modifiers, I'm ok if we decide to make it more C++-ish.

Reply:
Can't we do "Column type mismatch rhs=" + to_string(lhs->dtype) + " lhs=" + to_string(rhs->dtype)? Need to check this compiles but there must be a way to do this in a C++ way then relying on the old C style mechanism (this sometimes leads to a weird bug if type specifier and actual type mismatch). I need to check RAFT_EXPECTS internals but we may be able to make "Column type mismatch rhs=" << lhs->dtype << " lhs=" << lhs->dtype work as well to avoid to_string calls (if not already works).

Reply:
We may be able to reconcile, but consistency with cuDF is nice too. Let's continue the discussion in #11.